### PR TITLE
Timeouts

### LINKED ISSUE #91: No timeouts
As far as I can see I can't specify a timeout on the connection.  Not with the JdkRequest or with the ApacheRequest.

I implemented timeouts in the following fork, but I have some trouble with qulice before I can submit a pull request:
https://github.com/benhinssen/jcabi-http/tree/timeouts

### PR DESCRIPTION
**Pull request by benhinssen**
Fixes #91 

Added the possibility to specify a timeout on the UrlConnection / HttpRequest, both in the ApacheRequest and the JdkRequest.

Please advise on my changes.  If anything needs to be changed in order to be merged back, please let me know.

Thanks!

### CODE REVIEW

**Comment by dmarkov** [comment]
@benhinssen Thanks, let me find someone who can review this pull request

**Comment by dmarkov** [comment]
@pinaf review this one please

**Comment by benhinssen** [comment]
@dmarkov Thanks!

**Comment by pinaf** [comment]
@benhinssen there are a lot of checkstyle violations in the build logs. please fix those first.

**Comment by benhinssen** [comment]
@pinaf I made sure qulice return only one error, and that is naming convention.  I made sure one can set both the read as connect timeout.  So parameters are called readTimeout and connectTimeout.
One solution is to refactor it to readtimeout and connecttimeout, but that feels like cheating.
Another is to drop support for read and connect timeout separately, and just have a timeout field.

Please advise and I will refactor accordingly.

**Comment by pinaf** [comment]
@benhinssen is this related to an open issue? If so, please make a reference to it on the first comment :)

File: src/main/java/com/jcabi/http/request/ApacheRequest.java [hunk 1]
```diff
@@ -123,10 +130,12 @@ public String getMethod() {
                 };
             final URI uri = URI.create(home);
             req.setConfig(
-                RequestConfig.custom()
-                    .setCircularRedirectsAllowed(false)
-                    .setRedirectsEnabled(false)
-                    .build()
+                    RequestConfig.custom()
```
Comment (pinaf) [inline src/main/java/com/jcabi/http/request/ApacheRequest.java]: @benhinssen indentation is broken here (too much)

File: src/main/java/com/jcabi/http/Wire.java [hunk 2]
```diff
@@ -66,12 +66,15 @@
...
+     * @param connectTimeout The connect timeout
+     * @param readTimeout The read timeout
      * @return Response obtained
      * @throws IOException if fails
      * @checkstyle ParameterNumber (6 lines)
      */
     Response send(Request req, String home, String method,
-        Collection<Map.Entry<String, String>> headers, InputStream content)
+        Collection<Map.Entry<String, String>> headers, InputStream content,
+        int connectTimeout, int readTimeout)
```
Comment (pinaf) [inline src/main/java/com/jcabi/http/Wire.java]: @benhinssen `connect` and `read` are fine

File: src/main/java/com/jcabi/http/request/ApacheRequest.java [hunk 3]
```diff
@@ -87,10 +87,15 @@
         public Response send(final Request req, final String home,
             final String method,
             final Collection<Map.Entry<String, String>> headers,
-            final InputStream content) throws IOException {
+            final InputStream content,
+            final int connectTimeout,
```
Comment (pinaf) [inline src/main/java/com/jcabi/http/request/ApacheRequest.java]: @benhinssen `connect` and `read` are fine

File: src/main/java/com/jcabi/http/request/ApacheRequest.java [hunk 4]
```diff
@@ -113,7 +118,9 @@ public Response send(final Request req, final String home,
         public HttpEntityEnclosingRequestBase httpRequest(final String home,
             final String method,
             final Collection<Map.Entry<String, String>> headers,
-            final InputStream content) throws IOException {
+            final InputStream content,
+            final int connectTimeout,
```
Comment (pinaf) [inline src/main/java/com/jcabi/http/request/ApacheRequest.java]: @benhinssen `connect` and `read` are fine

Comment (pinaf) [inline src/main/java/com/jcabi/http/request/ApacheRequest.java]: @benhinssen please add the params to the javadoc

File: src/main/java/com/jcabi/http/request/BaseRequest.java [hunk 5]
```diff
@@ -107,6 +107,16 @@
     private final transient String mtd;
 
     /**
+     * Socket timeout to use.
+     */
+    private final transient int connectTimeout;
```
Comment (pinaf) [inline src/main/java/com/jcabi/http/request/BaseRequest.java]: @benhinssen `connect`

File: src/main/java/com/jcabi/http/request/BaseRequest.java [hunk 6]
```diff
@@ -107,6 +107,16 @@
...
+
+    /**
+     * Read timeout to use.
+     */
+    private final transient int readTimeout;
```
Comment (pinaf) [inline src/main/java/com/jcabi/http/request/BaseRequest.java]: @benhinssen `read`

File: src/main/java/com/jcabi/http/request/BaseRequest.java [hunk 7]
```diff
@@ -140,8 +150,26 @@
...
+     * @param body HTTP request body
+     * @param cnnctTimeout Connect timeout for http connection
+     * @param rdTimeout Read timeout for http connection
+     * @checkstyle ParameterNumber (5 lines)
+     */
+    BaseRequest(final Wire wre, final String uri,
         final Iterable<Map.Entry<String, String>> headers,
-        final String method, final byte[] body) {
+        final String method, final byte[] body,
+        final int cnnctTimeout, final int rdTimeout) {
```
Comment (pinaf) [inline src/main/java/com/jcabi/http/request/BaseRequest.java]: @benhinssen `cnct` and `rdd`

File: src/main/java/com/jcabi/http/request/BaseRequest.java [hunk 8]
```diff
@@ -140,8 +150,26 @@
      * @checkstyle ParameterNumber (5 lines)
      */
     BaseRequest(final Wire wre, final String uri,
+                final Iterable<Map.Entry<String, String>> headers,
```
Comment (pinaf) [inline src/main/java/com/jcabi/http/request/BaseRequest.java]: @benhinssen indentation

File: src/main/java/com/jcabi/http/request/BaseRequest.java [hunk 9]
```diff
@@ -140,8 +150,26 @@
...
+                final String method, final byte[] body) {
```
Comment (pinaf) [inline src/main/java/com/jcabi/http/request/BaseRequest.java]: @benhinssen indentation

File: src/main/java/com/jcabi/http/request/BaseRequest.java [hunk 10]
```diff
@@ -297,8 +342,9 @@ private Response fetchResponse(final InputStream stream)
         throws IOException {
         final long start = System.currentTimeMillis();
         final Response response = this.wire.send(
-            this, this.home, this.mtd,
-            this.hdrs, stream
+                this, this.home, this.mtd,
```
Comment (pinaf) [inline src/main/java/com/jcabi/http/request/BaseRequest.java]: @benhinssen indentation

File: src/main/java/com/jcabi/http/request/FakeRequest.java [hunk 11]
```diff
@@ -85,7 +85,9 @@
             public Response send(final Request req, final String home,
                 final String method,
                 final Collection<Map.Entry<String, String>> headers,
-                final InputStream text) throws IOException {
+                final InputStream text,
+                final int connectTimeout,
```
Comment (pinaf) [inline src/main/java/com/jcabi/http/request/FakeRequest.java]: @benhinssen `connect` and `read`

File: src/main/java/com/jcabi/http/request/JdkRequest.java [hunk 12]
```diff
@@ -84,7 +84,9 @@
         public Response send(final Request req, final String home,
             final String method,
             final Collection<Map.Entry<String, String>> headers,
-            final InputStream content) throws IOException {
+            final InputStream content,
+            final int connectTimeout,
```
Comment (pinaf) [inline src/main/java/com/jcabi/http/request/JdkRequest.java]: @benhinssen `connect` and `read`

File: src/main/java/com/jcabi/http/wire/AutoRedirectingWire.java [hunk 13]
```diff
@@ -107,9 +107,11 @@ public AutoRedirectingWire(final Wire wire, final int retries) {
     public Response send(final Request req, final String home,
         final String method,
         final Collection<Map.Entry<String, String>> headers,
-        final InputStream content) throws IOException {
+        final InputStream content,
+        final int connectTimeout,
```
Comment (pinaf) [inline src/main/java/com/jcabi/http/wire/AutoRedirectingWire.java]: @benhinssen `connect` and `read`

File: src/main/java/com/jcabi/http/wire/BasicAuthWire.java [hunk 14]
```diff
@@ -106,7 +106,9 @@ public BasicAuthWire(@NotNull(message = "wire can't be NULL")
     public Response send(final Request req, final String home,
         final String method,
         final Collection<Map.Entry<String, String>> headers,
-        final InputStream content) throws IOException {
+        final InputStream content,
+        final int connectTimeout,
```
Comment (pinaf) [inline src/main/java/com/jcabi/http/wire/BasicAuthWire.java]: @benhinssen `connect` and `read`

File: src/main/java/com/jcabi/http/wire/BasicAuthWire.java [hunk 15]
```diff
@@ -143,6 +145,8 @@ public Response send(final Request req, final String home,
                 throw new IllegalStateException(ex);
             }
         }
-        return this.origin.send(req, home, method, hdrs, content);
+        return this.origin.send(
+                req, home, method, hdrs, content, connectTimeout, readTimeout
```
Comment (pinaf) [inline src/main/java/com/jcabi/http/wire/BasicAuthWire.java]: @benhinssen too much indentation

File: src/main/java/com/jcabi/http/wire/CachingWire.java [hunk 16]
```diff
@@ -150,7 +150,9 @@ public CachingWire(
     public Response send(final Request req, final String home,
         final String method,
         final Collection<Map.Entry<String, String>> headers,
-        final InputStream content) throws IOException {
+        final InputStream content,
+        final int connectTimeout,
```
Comment (pinaf) [inline src/main/java/com/jcabi/http/wire/CachingWire.java]: @benhinssen `connect` and `read`

File: src/main/java/com/jcabi/http/wire/CachingWire.java [hunk 17]
```diff
@@ -169,14 +171,18 @@ public Response send(final Request req, final String home,
             try {
                 rsp = CachingWire.CACHE.get(this).get(
                     new CachingWire.Query(
-                        this.origin, req, home, headers, content
+                        this.origin, req, home, headers, content,
+                            connectTimeout, readTimeout
```
Comment (pinaf) [inline src/main/java/com/jcabi/http/wire/CachingWire.java]: @benhinssen indentation

File: src/main/java/com/jcabi/http/wire/CachingWire.java [hunk 18]
```diff
@@ -169,14 +171,18 @@ public Response send(final Request req, final String home,
...
                     )
                 );
             } catch (final ExecutionException ex) {
                 throw new IOException(ex);
             }
         } else {
-            rsp = this.origin.send(req, home, method, headers, content);
+            rsp = this.origin.send(
+                    req, home, method, headers, content,
```
Comment (pinaf) [inline src/main/java/com/jcabi/http/wire/CachingWire.java]: @benhinssen indentation

File: src/main/java/com/jcabi/http/wire/CachingWire.java [hunk 19]
```diff
@@ -208,22 +214,36 @@ public Response send(final Request req, final String home,
          */
         private final transient InputStream body;
         /**
+         * Connect timeout.
+         */
+        private final transient int connectTimeout;
```
Comment (pinaf) [inline src/main/java/com/jcabi/http/wire/CachingWire.java]: @benhinssen `connect`

File: src/main/java/com/jcabi/http/wire/CachingWire.java [hunk 20]
```diff
@@ -208,22 +214,36 @@ public Response send(final Request req, final String home,
...
+        /**
+         * Read timeout.
+         */
+        private final transient int readTimeout;
```
Comment (pinaf) [inline src/main/java/com/jcabi/http/wire/CachingWire.java]: @benhinssen `read`

File: src/main/java/com/jcabi/http/wire/CachingWire.java [hunk 21]
```diff
@@ -208,22 +214,36 @@ public Response send(final Request req, final String home,
...
          * @param hdrs Headers
          * @param input Input body
+         * @param cnnctTimeout Connect timeout
+         * @param rdTimeout Read timeout
          * @checkstyle ParameterNumberCheck (5 lines)
          */
         Query(final Wire wire, final Request req, final String home,
             final Collection<Map.Entry<String, String>> hdrs,
-            final InputStream input) {
+            final InputStream input, final int cnnctTimeout,
```
Comment (pinaf) [inline src/main/java/com/jcabi/http/wire/CachingWire.java]: @benhinssen `cnct` and `rdd`

File: src/main/java/com/jcabi/http/wire/CachingWire.java [hunk 22]
```diff
@@ -232,7 +252,8 @@ public Response send(final Request req, final String home,
          */
         public Response fetch() throws IOException {
             return this.origin.send(
-                this.request, this.uri, Request.GET, this.headers, this.body
+                this.request, this.uri, Request.GET, this.headers, this.body,
+                    this.connectTimeout, this.readTimeout
```
Comment (pinaf) [inline src/main/java/com/jcabi/http/wire/CachingWire.java]: @benhinssen indentation

File: src/main/java/com/jcabi/http/wire/CookieOptimizingWire.java [hunk 23]
```diff
@@ -97,7 +97,9 @@ public CookieOptimizingWire(@NotNull(message = "wire can't be NULL")
     public Response send(final Request req, final String home,
         final String method,
         final Collection<Map.Entry<String, String>> headers,
-        final InputStream content) throws IOException {
+        final InputStream content,
+        final int connectTimeout,
```
Comment (pinaf) [inline src/main/java/com/jcabi/http/wire/CookieOptimizingWire.java]: @benhinssen `connect` and `read`

File: src/main/java/com/jcabi/http/wire/OneMinuteWire.java [hunk 24]
```diff
@@ -87,7 +87,11 @@ public OneMinuteWire(@NotNull(message = "wire can't be NULL")
     public Response send(final Request req, final String home,
         final String method,
         final Collection<Map.Entry<String, String>> headers,
-        final InputStream content) throws IOException {
-        return this.origin.send(req, home, method, headers, content);
+        final InputStream content,
+        final int connectTimeout,
+        final int readTimeout) throws IOException {
+        return this.origin.send(
+                req, home, method, headers, content, connectTimeout, readTimeout
```
Comment (pinaf) [inline src/main/java/com/jcabi/http/wire/OneMinuteWire.java]: @benhinssen indentation

File: src/main/java/com/jcabi/http/wire/RetryWire.java [hunk 25]
```diff
@@ -89,7 +89,8 @@ public RetryWire(@NotNull(message = "wire can't be NULL")
     public Response send(final Request req, final String home,
         final String method,
         final Collection<Map.Entry<String, String>> headers,
-        final InputStream content) throws IOException {
+        final InputStream content,
+        final int connectTimeout, final int readTimeout) throws IOException {
```
Comment (pinaf) [inline src/main/java/com/jcabi/http/wire/RetryWire.java]: @benhinssen `connect` and `read`

File: src/main/java/com/jcabi/http/wire/RetryWire.java [hunk 26]
```diff
@@ -99,7 +100,8 @@ public Response send(final Request req, final String home,
             }
             try {
                 final Response rsp = this.origin.send(
-                    req, home, method, headers, content
+                    req, home, method, headers, content,
+                        connectTimeout, readTimeout
```
Comment (pinaf) [inline src/main/java/com/jcabi/http/wire/RetryWire.java]: @benhinssen indentation

File: src/main/java/com/jcabi/http/wire/TrustedWire.java [hunk 27]
```diff
@@ -108,7 +108,8 @@ public TrustedWire(final Wire wire) {
     public Response send(final Request req, final String home,
         final String method,
         final Collection<Map.Entry<String, String>> headers,
-        final InputStream content) throws IOException {
+        final InputStream content,
+        final int connectTimeout, final int readTimeout) throws IOException {
```
Comment (pinaf) [inline src/main/java/com/jcabi/http/wire/TrustedWire.java]: @benhinssen `connect` and `read`

File: src/main/java/com/jcabi/http/wire/TrustedWire.java [hunk 28]
```diff
@@ -116,7 +117,10 @@ public Response send(final Request req, final String home,
                 HttpsURLConnection.setDefaultSSLSocketFactory(
                     TrustedWire.context().getSocketFactory()
                 );
-                return this.origin.send(req, home, method, headers, content);
+                return this.origin.send(
+                        req, home, method, headers, content,
```
Comment (pinaf) [inline src/main/java/com/jcabi/http/wire/TrustedWire.java]: @benhinssen indentation

File: src/main/java/com/jcabi/http/wire/UserAgentWire.java [hunk 29]
```diff
@@ -103,7 +103,9 @@ public UserAgentWire(@NotNull(message = "wire can't be NULL")
     public Response send(final Request req, final String home,
         final String method,
         final Collection<Map.Entry<String, String>> headers,
-        final InputStream content) throws IOException {
+        final InputStream content,
+        final int connectTimeout,
```
Comment (pinaf) [inline src/main/java/com/jcabi/http/wire/UserAgentWire.java]: @benhinssen `connect` and `read`

File: src/main/java/com/jcabi/http/wire/UserAgentWire.java [hunk 30]
```diff
@@ -121,6 +123,8 @@ public Response send(final Request req, final String home,
                 )
             );
         }
-        return this.origin.send(req, home, method, hdrs, content);
+        return this.origin.send(
+                req, home, method, hdrs, content, connectTimeout, readTimeout
```
Comment (pinaf) [inline src/main/java/com/jcabi/http/wire/UserAgentWire.java]: @benhinssen indentation

File: src/main/java/com/jcabi/http/wire/VerboseWire.java [hunk 31]
```diff
@@ -88,7 +88,9 @@ public VerboseWire(@NotNull(message = "wire can't be NULL")
     public Response send(final Request req, final String home,
         final String method,
         final Collection<Map.Entry<String, String>> headers,
-        final InputStream content) throws IOException {
+        final InputStream content,
+        final int connectTimeout,
```
Comment (pinaf) [inline src/main/java/com/jcabi/http/wire/VerboseWire.java]: @benhinssen `connect` and `read`

**Comment by pinaf** [comment]
@benhinssen 32 comments above.

**Comment by pinaf** [comment]
@benhinssen let's get travis to show a green light :)

**Comment by pinaf** [comment]
@benhinssen mostly variable names and indentation

**Comment by benhinssen** [comment]
@pinaf this is not related to any open issue.  I made this change because I needed it in the project I was using this library.  I wanted to share it back with you guys.
I'll make the changes and make sure travis gives a green light!

Thanks!

**Comment by pinaf** [comment]
@benhinssen got it. thanks!

**Comment by benhinssen** [comment]
@pinaf Now travis is complaining about code I didn't change. :-)

**Comment by pinaf** [comment]
@benhinssen The problem is a duplicated string. line 124 is just the first place where it appears.
>
[ERROR] /src/main/java/com/jcabi/http/Request.java[124-124]: The String literal "request is never NULL" appears 4 times in this file; the first occurrence is on line 124 (AvoidDuplicateLiterals)

try extracting that repeated string to a field.
```java
private static final String NOTNULL = "request is never NULL";
```

and reference it with `Request.NOTNULL` in the 4 different places.

**Comment by benhinssen** [comment]
@pinaf indeed, I added one occurrence which triggers the rule.

**Comment by benhinssen** [comment]
@pinaf private constants are not allowed in an interface, so I made a normal field.

**Comment by pinaf** [comment]
@benhinssen crap. the problem is that it will be public which is a no-no... let's revert that and add a suppression to the PMD error.

**Comment by pinaf** [comment]
@benhinssen please add
```java
@SupressWarnings("PMD.AvoidDuplicateLiterals")
```
above the interface declaration.

**Comment by benhinssen** [comment]
@pinaf It would indeed be confusing to use this as a public field, then it looks like part of the api, which is not.

**Comment by benhinssen** [comment]
@pinaf now it should be ok.

File: src/main/java/com/jcabi/http/request/ApacheRequest.java [hunk 32]
```diff
@@ -124,9 +137,11 @@ public String getMethod() {
             final URI uri = URI.create(home);
             req.setConfig(
                 RequestConfig.custom()
-                    .setCircularRedirectsAllowed(false)
-                    .setRedirectsEnabled(false)
-                    .build()
+                        .setCircularRedirectsAllowed(false)
```
Comment (pinaf) [inline src/main/java/com/jcabi/http/request/ApacheRequest.java]: @benhinssen can't we move these lines back a bit? The old alignment was correct.

File: src/main/java/com/jcabi/http/request/BaseRequest.java [hunk 33]
```diff
@@ -276,9 +321,9 @@ public String toString() {
         for (final Map.Entry<String, String> header : this.hdrs) {
             text.append(
                 Logger.format(
-                    "%s: %s\n",
-                    header.getKey(),
-                    header.getValue()
+                        "%s: %s\n",
```
Comment (pinaf) [inline src/main/java/com/jcabi/http/request/BaseRequest.java]: @benhinssen let's revert this identation change

File: src/main/java/com/jcabi/http/wire/CookieOptimizingWire.java [hunk 34]
```diff
@@ -137,6 +139,8 @@ public Response send(final Request req, final String home,
                 )
             );
         }
-        return this.origin.send(req, home, method, hdrs, content);
+        return this.origin.send(
+                req, home, method, hdrs, content, connect, read
```
Comment (pinaf) [inline src/main/java/com/jcabi/http/wire/CookieOptimizingWire.java]: @benhinssen mode indentation back a bit.

File: src/main/java/com/jcabi/http/request/ApacheRequest.java [hunk 35]
```diff
@@ -87,10 +87,15 @@
...
-            final InputStream content) throws IOException {
+            final InputStream content,
+            final int connect,
+            final int read) throws IOException {
             final CloseableHttpResponse response =
                 HttpClients.createSystem().execute(
-                    this.httpRequest(home, method, headers, content)
+                    this.httpRequest(
+                        home, method, headers, content,
+                            connect, read
```
Comment (pinaf) [inline src/main/java/com/jcabi/http/request/ApacheRequest.java]: @benhinssen align this to the line above: `home`

**Comment by pinaf** [comment]
@benhinssen a few comments on indentation above

**Comment by benhinssen** [comment]
@pinaf Sorry, indeed, you are right, I didn't change this intentionally.  I should really find the setting in intellij causing this.

**Comment by pinaf** [comment]
@benhinssen much better :)

**Comment by pinaf** [comment]
@rultor merge

**Comment by benhinssen** [comment]
@pinaf I forgot that I created an issue for this before changing it myself.  This was issue #91.  I referenced it in the first comment as you requested earlier.

**Comment by pinaf** [comment]
@benhinssen great! thanks!

**Comment by yegor256** [comment]
@rultor merge

**Comment by benhinssen** [comment]
@yegor256 Thanks!

**Comment by cvrebert** [comment]
@yegor256 Will there be a new release with this soon?

**Comment by yegor256** [comment]
@rultor release, tag is `1.11`

**Comment by cvrebert** [comment]
> Error creating blob: You have triggered an abuse detection mechanism. Please wait a few minutes before you try again. (403)

**Comment by yegor256** [comment]
@rultor release, tag is `1.11`

**Comment by dmarkov** [comment]
@pinaf I added **28 mins** to your account, many thanks for working with the project! The completion time here was AP-26D81886G9735604Y.; you're getting extra minutes here (c=13); added +28 to your rating, now it is equal to [+6643](http://www.netbout.com/b/35128?open=rating)

**Comment by dmarkov** [comment]
@rultor deploy pls

**Pull request merged**